src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpWindowsProxy.cs

```diff
@@ -6,45 +6,70 @@
-        private readonly List<IPAddress>? _localIp;
+        private readonly RegistryKey? _internetSettingsRegistry = Registry.CurrentUser?.OpenSubKey("Software\\Microsoft\\Windows\\CurrentVersion\\Internet Settings");
+        private MultiProxy _insecureProxy;      // URI of the http system proxy if set
+        private  MultiProxy _secureProxy;       // URI of the https system proxy if set
+        private FailedProxyCache _failedProxies = new FailedProxyCache();
+        private List<string>? _bypass;          // list of domains not to proxy
+        private List<IPAddress>? _localIp;
         private ICredentials? _credentials;
-        private readonly WinInetProxyHelper _proxyHelper;
+        private WinInetProxyHelper _proxyHelper;
         private SafeWinHttpHandle? _sessionHandle;
         private bool _disposed;
+        private EventWaitHandle _waitHandle = new EventWaitHandle(false, EventResetMode.AutoReset);
+        private const int _registrationFlags = Interop.Advapi32.REG_NOTIFY_CHANGE_NAME | Interop.Advapi32.REG_NOTIFY_CHANGE_LAST_SET | Interop.Advapi32.REG_NOTIFY_CHANGE_ATTRIBUTES | Interop.Advapi32.REG_NOTIFY_THREAD_AGNOSTIC;
+        private RegisteredWaitHandle? _registeredWaitHandle;
 
         public static bool TryCreate([NotNullWhen(true)] out IWebProxy? proxy)
         {
-            // This will get basic proxy setting from system using existing
-            // WinInetProxyHelper functions. If no proxy is enabled, it will return null.
-            SafeWinHttpHandle? sessionHandle = null;
-            proxy = null;
+            proxy = new HttpWindowsProxy();
+            return true;
+        }
 
-            WinInetProxyHelper proxyHelper = new WinInetProxyHelper();
-            if (!proxyHelper.ManualSettingsOnly && !proxyHelper.AutoSettingsUsed)
+        public HttpWindowsProxy()
+        {
+
+            if (_internetSettingsRegistry != null)
             {
-                return false;
+                // we register for change notifications so we can react to changes during lifetime.
+                if (Interop.Advapi32.RegNotifyChangeKeyValue(_internetSettingsRegistry.Handle, true, _registrationFlags, _waitHandle.SafeWaitHandle, true) == 0)
+                {
+                    _registeredWaitHandle = ThreadPool.RegisterWaitForSingleObject(_waitHandle, RegistryChangeNotificationCallback, this, -1, false);
+                }
             }
 
+            _proxyHelper = new WinInetProxyHelper();
+            UpdateConfiguration();
+        }
+
+        private static void RegistryChangeNotificationCallback(object? state, bool timedOut)
+        {
+            HttpWindowsProxy proxy = (HttpWindowsProxy)state!;
+            // We need to register for notification every time
+            Interop.Advapi32.RegNotifyChangeKeyValue(proxy._internetSettingsRegistry!.Handle, true, _registrationFlags, proxy._waitHandle.SafeWaitHandle, true);
```

Comment:
Is it possible this instance gets Disposed of concurrently with a change notification? Assuming yes, should we be dealing with that race condition here, even if just catching an ObjectDisposedException?

Reply:
I added guard for it.